sklearn/metrics/cluster/supervised.py

```diff
@@ -814,8 +816,82 @@ def normalized_mutual_info_score(labels_true, labels_pred, max_n_classes=5000):
     return nmi
 
 
+def fowlkes_mallows_score(labels_true, labels_pred, max_n_classes=5000):
+    r"""Mesure the similarity of two clusterings of a set of points.
+
+    The Fowlkes-Mallows index (FMI) is defined by the following formula::
+
+        FMI = TP / sqrt((TP + FP) * (TP + FN))
+
+    With :math:`TP` is the number of pair of points that belongs in the same
+    clusters in both labels_true and labels_pred, :math:`FP` is the number
+    of pair of points that belongs in the same clusters in labels_true and not
+    in labels_pred and :math:`FN`is the number of pair of points that belongs
+    in the same clusters in labels_pred and not in labels_True
+
+    The score ranges from 0 to 1. A high value indicates a good similarity
+    between two clusters.
+
+    Parameters
+    ----------
+    labels_true : int array, shape = [n_samples]
+        A clustering of the data into disjoint subsets.
+
+    labels_pred : array, shape = [n_samples]
+        A clustering of the data into disjoint subsets.
+
+    max_n_classes: int, optional (default=5000)
+        Maximal number of classes handled by the adjusted_rand_score
+        metric. Setting it too high can lead to MemoryError or OS
+        freeze
+
+    Returns
+    -------
+    score: float
+       The resulting Fowlkes-Mallows score.
+
+    Examples
+    --------
+
+    Perfect labelings are both homogeneous and complete, hence have
+    score 1.0::
+
+      >>> from sklearn.metrics.cluster import fowlkes_mallows_score
+      >>> fowlkes_mallows_score([0, 0, 1, 1], [0, 0, 1, 1])
+      1.0
+      >>> fowlkes_mallows_score([0, 0, 1, 1], [1, 1, 0, 0])
+      1.0
+
+    If classes members are completely split across different clusters,
+    the assignment is totally in-complete, hence the FMI is null::
+
+      >>> fowlkes_mallows_score([0, 0, 0, 0], [0, 1, 2, 3])
+      0.0
+
+    References
+    ----------
+    .. [1] `E. B. Fowkles and C. L. Mallows, 1983. "A method for comparing two
+       hierarchical clusterings". Journal of the American Statistical
+       Association
+       <http://wildfire.stat.ucla.edu/pdflibrary/fowlkes.pdf>`_
+
+    .. [2] `Wikipedia entry for the Fowlkes-Mallows Index
+           <https://en.wikipedia.org/wiki/Fowlkes-Mallows_index>`_
+    """
+    labels_true, labels_pred = check_clusterings(labels_true, labels_pred,)
+    n_samples, = labels_true.shape
+
+    c = contingency_matrix(labels_true, labels_pred,
+                           max_n_classes=max_n_classes)
+    tk = np.sum(c ** 2) - n_samples
```

Comment:
Can use `np.dot(c.ravel(), c.ravel())` or similar, though I guess there are only real efficiency gains if the number of clusters is large.